### PR TITLE
chore: `Dockerfile` remove redundant directives

### PR DESCRIPTION
**Pull request by polarathene**
- Removes the `/etc/nsswitch.conf` workaround.
- Removes the redundant usage of `VOLUME` directives.

## Related issue(s)

The associated issues have been ignored for over a year. They've now been marked as stale, this PR attempts to address them for this repo.

- https://github.com/ory/hydra/issues/3685
- https://github.com/ory/hydra/issues/3683

See the issues for detailed justification of the summarized changes.

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [ ] I am following the [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

These changes should be rather straight-forward. The bulk of the changeset is just noise repeating the same lines across several Dockerfiles, but I did notice inconsistencies which hint that these files may need to be revisited by someone more familiar with them.

Especially with the HSM image which appears to have had the `runner` stage broken since [this June 2023 PR](https://github.com/ory/hydra/pull/3539). though only earlier stages are built with the Makefile:

https://github.com/ory/hydra/blob/8e71f9178d9a21dead8a260bae60b8406710cf5b/Makefile#L89-L96

The HSM quickstart compose example should attempt to build the `runner` stage and fail:

https://github.com/ory/hydra/blob/8e71f9178d9a21dead8a260bae60b8406710cf5b/quickstart-hsm.yml#L16

### CODE REVIEW

**Comment by polarathene** [comment]
Given the past associated issues were neglected, I am not too keen to invest more time than necessary.

I'll give the CLA a look and the other checklist items if this PR actually gets acknowledged with interest to merge it.

---

The scratch image seems redundant, you've got a [replacement image](https://github.com/ory/hydra/blob/8e71f9178d9a21dead8a260bae60b8406710cf5b/.docker/Dockerfile-distroless-static) now with the rough equivalent via Google distroless base image. You might as well drop it? The main difference apart from noted issues in the scratch Dockerfile is also a lack of the sqlite support.

Your alpine vs sqlite (alpine) images are effectively the same too, except for the `sqlite` package and slightly different default CMD. Probably no need or benefit maintaining the two with such a low distinction?

File: .docker/Dockerfile-alpine [hunk 1]
```diff
@@ -1,22 +1,25 @@
...
+      --shell /sbin/nologin \
+      ory
 
-# By creating the sqlite folder as the ory user, the mounted volume will be owned by ory:ory, which
-# is required for read/write of SQLite.
-RUN mkdir -p /var/lib/sqlite && \
-    chown ory:ory /var/lib/sqlite
+    # Create the sqlite directory with ownership to that user and group:
+    # NOTE: This is required for read/write by SQLite.
+    install --owner ory --group ory --directory /var/lib/sqlite
```
Comment (aeneasr) [inline .docker/Dockerfile-alpine:18]: What does `install` do here? I've never seen this command

File: .docker/Dockerfile-hsm [hunk 2]
```diff
@@ -53,30 +53,36 @@ ENV HSM_LIBRARY=/usr/lib/softhsm/libsofthsm2.so
 ENV HSM_TOKEN_LABEL=hydra
 ENV HSM_PIN=1234
 
+# NOTE: This is broken already. Even though this image provides a shell, you'd need to configure it with
+# `SHELL ["/busybox/sh", "-c"]`, however `apt-get` does not exist either in a distroless image.
+# This was original an Alpine image, the refactoring was not verified properly in this commit:
+# https://github.com/ory/hydra/commit/c1e1a569621d88365dceee7372ca49ecd119f939#diff-ae54bef08e3587b28ad8e93eb253a9a5cd9ea6f4251977e35b88dc6b42329e25L31
```
Comment (aeneasr) [inline .docker/Dockerfile-hsm:59]: The HSM image is really just to run some e2e hsm tests. It's not being distributed and should not be used.

**Review by aeneasr — COMMENTED** [review]
> Given the past associated issues were neglected, I am not too keen to invest more time than necessary.

No problem!

> The scratch image seems redundant, you've got a [replacement image](https://github.com/ory/hydra/blob/8e71f9178d9a21dead8a260bae60b8406710cf5b/.docker/Dockerfile-distroless-static) now with the rough equivalent via Google distroless base image. You might as well drop it? The main difference apart from noted issues in the scratch Dockerfile is also a lack of the sqlite support.

Some dockerfiles are only used in the repo itself (for dev purposes) and some are being pushed to our docker registry. You can find the files used for prod distribution in the goreleaser config. Generally not too keen to deprecate image variants because someone always complains about it.

> Your alpine vs sqlite (alpine) images are effectively the same too, except for the sqlite package and slightly different default CMD. Probably no need or benefit maintaining the two with such a low distinction?

That makes sense but please in another pr

File: .docker/Dockerfile-hsm (hunk 2 above)
Comment (aeneasr) [reply .docker/Dockerfile-hsm:59]: Yeah, e2e tests are now failing. Probably just revert the changes here:  https://github.com/ory/hydra/actions/runs/12523819546/job/34933781514?pr=3914

**Comment by polarathene** [comment]
> Generally not too keen to deprecate image variants because someone always complains about it.

Since they're the same though, just use the same `Dockerfile` and publish it as two separate images? That should avoid your concern about dropping the original source file for an image variant?

> > Your alpine vs sqlite (alpine) images are effectively the same too, except for the sqlite package and slightly different default CMD. Probably no need or benefit maintaining the two with such a low distinction?
> 
> That makes sense but please in another pr

👍 

---

> The HSM image is really just to run some e2e hsm tests. It's not being distributed and should not be used.

No worries, just pointing out that it's presently broken since the switch, just AFAIK your tests don't use this final stage of the image so it's not getting caught by them.

> Yeah, e2e tests are now failing. Probably just revert the changes here
> 
> https://github.com/ory/hydra/actions/runs/12523819546/job/34933781514?pr=3914

That is an easy fix, it just wants `apt-get` instead of `apt`.

**Comment by polarathene** [comment]
**TL;DR:** Spotted a few more bugs/issues while looking into responding to your feedback.

## Registries and tags (publishing bug)

> Some dockerfiles are only used in the repo itself (for dev purposes) and some are being pushed to our docker registry.

Is the DockerHub registry `oryd/hydra` no longer being used? (_**EDIT:** Yes it is, just no new hydra release for approx a year, redundant images are not actually published as implied_)
- It has some tags for images as SHA256 digests with `.sig` extension that don't seem to be actual images, those shouldn't be there.
- The last release was about a year ago with [`latest` and `v2.2.0` tags published](https://hub.docker.com/r/oryd/hydra/tags). That seems to be the latest release on GH releases page too, so it seems like DockerHub is still a valid registry being maintained.
- While I do see these images published as multi-platform images, I'm also seeing individual images published with their platform as part of the tag (_eg: [`v2.2.0-arm64`](https://hub.docker.com/layers/oryd/hydra/v2.2.0-arm64/images/sha256-43b36e4d1baaed7707d9c353a4e7a22af35fbb66625d148651515d4785d8ffec) + [`v2.2.0-arm64-distroless`](https://hub.docker.com/layers/oryd/hydra/v2.2.0-arm64-distroless/images/sha256-721bc98224f28581204c1cfcd9d8f842a2e472ec69499d6c0c8e3cb7975f60c8)_) which seems redundant?

Perhaps something is a bit iffy with your release process?

> You can find the files used for prod distribution in the goreleaser config.

I'm not too familiar with GoReleaser, but I am with Github Actions equivalent for building and publishing images to DockerHub.

I only see two variables defined here:

https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/.goreleaser.yml#L11-L12

Which seems to align with your two variants on DockerHub `-distroless` tag suffix for Debian image with glibc and no package manager, while the version only tags are Alpine with musl and it's package manager.

I haven't done a thorough search for where the other Dockerfile variants are potentially being used (_which may be internal for you, or referenced externally from this repo like in docs?_), but unless there's another registry you had in mind it doesn't seem like your earlier concern about deprecating variants is applicable if they were never published to registries?

Your [install docs for Hydra with Docker](https://www.ory.sh/docs/hydra/self-hosted/install#docker) reference `oryd/hydra` but not a specific image registry, so that often defaults to DockerHub:

> image

No mention there of the individual platform variants or the `.sig` digest tags. I'm going to assume that's a publishing bug with your GoReleaser config?

## `oryd/hydra` default Alpine image vs `-sqlite` variant

Both this repo's [Get Started](https://github.com/ory/hydra/tree/master#get-started) and [Develop](https://github.com/ory/hydra/tree/master#build-docker) sections refer to a different Hydra docs page for Docker [which uses the `quickstart.yml`](https://www.ory.sh/docs/hydra/5min-tutorial), and that likewise refers to `oryd/hydra`:

https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/quickstart.yml#L14

Although the Develop section instructs `make docker` to build an image of the same tag locally instead of pulling from the registry, and that uses the `-sqlite` image variant `Dockerfile`:

https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/Makefile#L71-L74

Which as we have discussed earlier `-sqlite` variant is effectively the Alpine image with the `sqlite` package (_2.5 MB added weight uncompressed_) and [with `CMD` set to `serve`](https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/.docker/Dockerfile-sqlite#L36) which will just output help message as it [expects `serve all` / `serve public` / `serve admin`](https://www.ory.sh/docs/hydra/cli/hydra-serve), meanwhile the published images default to `serve all` and this `-sqlite` image if built locally as `oryd/hydra` has the container command modified in `quickstart.yml`... so there's really no compelling reason to keep this `Dockerfile` around either.

Technically though, the `quickstart.yml` has a specific release tag (good), while the `make docker` command defaults `IMAGE_TAG` to `latest` if not specified, and since the quickstart doesn't use that ENV, nor does the README instructions mention it in the [Develop section](https://github.com/ory/hydra/tree/master?tab=readme-ov-file#build-docker), the image is built locally but the compose config still pulls from DockerHub 😅 So that's another "bug" observed.

---

## Reference - Publishing image via GHA instead

For reference [this is what publishing to image registries](https://github.com/docker-mailserver/docker-mailserver/blob/24fb65ce7b638b342c1c9f427912756a0e53ed3e/.github/workflows/generic_publish.yml#L14-L81) looks like with Github Actions:

```yaml
# Tag an image with semver pattern (useful for tools that understand the tag convention),
# optionally with a tag rule that reflects your primary git branch builds for testing:
- name: 'Prepare tags'
  id: prep
  uses: docker/metadata-action@v5.6.1
  with:
    images: |
      ${{ secrets.DOCKER_REPOSITORY }}
      ${{ secrets.GHCR_REPOSITORY }}
    tags: |
      type=edge,branch=master
      type=semver,pattern={{major}}
      type=semver,pattern={{major}}.{{minor}}
      type=semver,pattern={{major}}.{{minor}}.{{patch}}


# Two registries to publish to configured (DockerHub and GHCR):
- name: 'Login to DockerHub'
  uses: docker/login-action@v3
  with:
    username: ${{ secrets.DOCKER_USERNAME }}
    password: ${{ secrets.DOCKER_PASSWORD }}
- name: 'Login to GitHub Container Registry'
  uses: docker/login-action@v3
  with:
    registry: ghcr.io
    username: ${{ github.actor }}
    password: ${{ secrets.GITHUB_TOKEN }}


# Build the image from a Dockerfile for the requested platforms,
# then publish a multi-platform image at the registries with the tags declared above
- name: 'Build and publish images'
  uses: docker/build-push-action@v6.10.0
  with:
    context: .
    # Customize any Dockerfile ARG values here:
    build-args: |
      DMS_RELEASE=${{ github.ref_type == 'tag' && github.ref_name || 'edge' }}
      VCS_REVISION=${{ github.sha }}
    platforms: linux/amd64,linux/arm64
    push: true
    tags: ${{ steps.prep.outputs.tags }}
    # Disable provenance attestation: https://docs.docker.com/build/attestations/slsa-provenance/
    provenance: false
```

That's all there is to it really, works well! (_I omitted a little bit from the link referenced for brevity_)

Github Actions also has reusable workflows, allowing you to setup a template with the above steps and pass some input variables in to support the different `Dockerfile` and tag variants for example. You can [see an example of that here](https://github.com/docker-mailserver/docker-mailserver/blob/24fb65ce7b638b342c1c9f427912756a0e53ed3e/.github/workflows/default_on_push.yml#L33-L39).

Perhaps GoReleaser is simpler than that, but the reference I've provided with official Github Actions from Docker doesn't have that same publishing weirdness at DockerHub for `mailserver/docker-mailserver` tags pushed.

**Comment by polarathene** [comment]
Documenting some more feedback should it appeal to any maintainers that want to pursue it.

## A `Dockerfile` with flexible sourcing of the Hydra binary

I only mention this since your currently maintaining both approaches via separate Dockerfiles.

[`Dockerfile-distroless-static`](https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/.docker/Dockerfile-distroless-static) is effectively the last stage of [`Dockerfile-build`](https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/.docker/Dockerfile-build), but without the `/var/lib/sqlite` directory.

If you want to have a `Dockerfile` that can build an image with the `hydra` binary `COPY` supporting both the `builder` stage, or a copy from context (supplied externally), you could have this with in the same `Dockerfile` as different stages that you select by a build target option instead of specifying different files.

```Dockerfile
FROM runner AS provide-hydra
COPY hydra /usr/bin/hydra

FROM runner AS build-hydra
COPY --from=builder /usr/bin/hydra /usr/bin/hydra
```

As the last stage will be the default target, it would build Hydra with the image build by default. To use the external build of Hydra instead just use:

```bash
docker build --target provide-hydra --file Dockerfile-build --tag oryd/hydra:latest-distroless .
```

### A `Dockerfile` that builds with configurable Alpine/Distroless base image

The `FROM runner` stage reference could be an ARG to switch between the Alpine an Debian supported image bases, which would also remove the separate `Dockerfile-alpine`.

Roughly, after the `builder` stage in `Dockerfile-build`, you could have the following:

```Dockerfile
# NOTE: ARG referenced via FROM must be defined at the top of the Dockerfile
# Place this above the builder stage
ARG BASE_IMAGE=alpine


# These stages setup the Alpine and Debian base images for the runtime stage:
FROM alpine:3.20 AS alpine
RUN apk add --no-cache ca-certificates
FROM gcr.io/distroless/static-debian12:nonroot AS distroless


# The runtime stage defaults to Alpine, to build the distroless variant instead use:
# `docker build --build-arg BASE_IMAGE=distroless`
FROM ${BASE_IMAGE} AS runtime
RUN --mount=from=busybox:latest,src=/bin/,dst=/bin/ mkdir -p /var/lib/sqlite
ENTRYPOINT ["hydra"]
CMD ["serve", "all"]


# Final stage decides where to retrieve the hydra binary from.
# Defaults to `build-hydra`, provide your own build via
# `docker build --target provide-hydra`
FROM runner AS provide-hydra
COPY hydra /usr/bin/hydra
FROM runner AS build-hydra
COPY --from=builder /usr/bin/hydra /usr/bin/hydra
```

All 4 image variants can then be built:

```bash
docker build --build-arg BASE_IMAGE=alpine --target build-hydra
docker build --build-arg BASE_IMAGE=alpine --target provide-hydra
docker build --build-arg BASE_IMAGE=distroless --target build-hydra
docker build --build-arg BASE_IMAGE=distroless --target provide-hydra
```

Personally though, if you're doing static builds of Hydra outside of the `Dockerfile` to copy into the image, if this is only for some runtime purpose like development, you could just use a volume mount instead?

If the `builder` stage were always used for copying the hydra binary, then the prior/existing `/var/lib/sqlite` dir creation approach is valid (_instead of the busybox approach to support `mkdir` with distroless_):

```Dockerfile
COPY --from=builder /var/lib/sqlite /var/lib/sqlite
```

You can also use a `FROM scratch` stage if you'd want the builder to output the binary to the host instead of an image via [`--output`](https://docs.docker.com/reference/cli/docker/buildx/build/#output):

```Dockerfile
FROM scratch AS export-hydra
COPY --from=builder /usr/bin/hydra /hydra
```

```bash
# The image contents is exported to the local filesystem:
docker buildx build --output ./ --target export-hydra --file Dockerfile-build .
```

That'd give you the inverse of copying an external build into the image if you found it more convenient to build with Docker and export the binary for releases elsewhere.

### `Dockerfile-hsm` improvements

Likewise for [`Dockerfile-hsm`](https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/.docker/Dockerfile-hsm), the `-tags` for Go need to be adjusted/appended but otherwise differences are minimal:

```Dockerfile
# Use the same `builder` stage from `Dockerfile-build`
# NOTE: this might require `go test` to also include the arg `-ldflags="-extldflags=-static"`?
# Modify the `builder` stage likewise use the `GO_BUILD_TAGS` ARG too
# Define the ARG above the builder stage to assign a common default
ARG GO_BUILD_TAGS=sqlite,sqlite_omit_load_extension


FROM builder AS test-hsm-base
ENV HSM_ENABLED=true
ENV HSM_LIBRARY=/usr/lib/softhsm/libsofthsm2.so
ENV HSM_TOKEN_LABEL=hydra
ENV HSM_PIN=1234

RUN apt-get -y install softhsm2 opensc
RUN pkcs11-tool --module "$HSM_LIBRARY" --slot 0 --init-token --so-pin 0000 --init-pin --pin "$HSM_PIN" --label "$HSM_TOKEN_LABEL"

FROM test-hsm-base AS test-hsm
ARG GO_BUILD_TAGS
RUN go test -p 1 -failfast -short -tags=${GO_BUILD_TAGS},hsm ./...

FROM test-hsm-base AS test-hsm-refresh
ARG GO_BUILD_TAGS UPDATE_SNAPSHOTS=true
RUN go test -p 1 -failfast -short -tags=${GO_BUILD_TAGS},hsm,refresh ./...
```

Much smaller/simpler to maintain like that.

The `GO_BUILD_TAGS` ARG could be avoided, but it might help to centralize on it, plus you can easily adjust them at build time for all image variants this way.

```bash
# Now you can build via target:
docker build --target test-hsm
docker build --target test-hsm-refresh

# And for the final stage of `Dockerfile-hsm` (but actually works),
# just set the build arg to that HSM stage:
docker build --build-arg BASE_IMAGE=test-hsm-base
```

While this could allow for a single `Dockerfile` for every image this repo builds, I don't really encourage that. Since the HSM images are only for internal use, you could just perform an extra build command to get the `builder` stage built as a separate image, before the `Dockerfile-hsm` references that via `FROM` (_or you could leverage [Bake config](https://docs.docker.com/build/bake/reference/#use-another-target-as-base) file to keep it as a single build command_).

Regardless, `Dockerfile-hsm` could be simplified with the above approach. You could also throw away the final stage since it doesn't actually build and apparently isn't used anywhere anyway?

**Review by polarathene — COMMENTED** [review]
Apologies about all the verbosity, I don't really want to think about all this again and move onto other things instead 😂 

Here's some final feedback to consider. I understand all my input could be overwhelming so I'll wrap this up with a checklist summary and links in a follow-up comment that should help make this more digestible for anyone interested.

**Comment by polarathene** [comment]
## Overview of everything above

This comment should summarize everything with links for more details if you need them.

## Changes from this PR

This PR applies the following changes:
- [x] Refactor Alpine images to be consistent with directives and common shared logic.
  - `RUN` directive now uses [HereDocs](https://docs.docker.com/reference/dockerfile/#here-documents) feature (_requires Docker v23 or `DOCKER_BUILDKIT=1` if older docker engine_).
  - 😎 Install `ca-certificates` + create `ory` user+group (_with UID/GID pinned for consistency_)
  - 😎 `/var/lib/sqlite` directory created via `install` command (_effectively `mkdir` + `chown`_)
  - 🚀 Remove legacy `nsswitch.conf` workaround
  - 🚀 Remove legacy `VOLUME` directive
  - 👀 Realization that `Dockefile-scratch` + `Dockerfile-sqlite` are redundant and could be deleted in future.
  - ⚠️ Realization that final stage of `Dockerfile-hsm` is broken, but stage isn't used (_removal in future proposed, see TODOs below for more info_).
- [x] `USER` directive shifted to end of each `Dockerfile`. Preferring non-root likely causes more issues than benefits (_removal in future proposed, see TODOs below for more info_).
- [x] Additional commentary. Some of it could be dropped, but adding directly into the source should see that it gets addressed more promptly or future maintenance / readers have better context to reference.

---

Overall if we strip the commentary and ignore minor differences, `Dockerfile-alpine`, `Dockerfile-scratch`, `Dockerfile-sqlite` and `Dockerfile-hsm` are all normalized to roughly this same common snippet:

```Dockerfile
RUN <<HEREDOC
    apk upgrade --no-cache
    apk add --no-cache --upgrade ca-certificates

    addgroup --system --gid 500 ory
    adduser --system --uid 500 \
      --gecos "Ory User" \
      --home /home/ory \
      --ingroup ory \
      --shell /sbin/nologin \
      ory

    install --owner ory --group ory --directory /var/lib/sqlite
HEREDOC

COPY hydra /usr/bin/hydra
ENTRYPOINT ["hydra"]
CMD ["serve", "all"]
USER ory
```

While addressing the two referenced issues at the start of the PR description (_removal of legacy `nsswitch.conf` and `VOLUME` directive_).

In future if dropping the non-root `ory` user, the Alpine and Distroless images effectively become this:

```Dockerfile
FROM alpine:3.20
RUN apk add --no-cache ca-certificates && mkdir -p /var/lib/sqlite

COPY hydra /usr/bin/hydra
ENTRYPOINT ["hydra"]
CMD ["serve", "all"]
```

```Dockerfile
FROM gcr.io/distroless/static-debian12:nonroot
RUN --mount=from=busybox:latest,src=/bin/,dst=/bin/ mkdir -p /var/lib/sqlite

COPY hydra /usr/bin/hydra
ENTRYPOINT ["hydra"]
CMD ["serve", "all"]
```

All the verbosity I've provided is towards instilling confidence to pursue that change via a follow-up PR.

## Additional context

- The DockerHub registry only publishes two images, Alpine (default) + Debian (`-distroless` suffix). HSM variant is only for testing. Each Dockerfile has `COPY` for the hydra externally, except for `Dockerfile-build` (_although this does [not appear to be used for publishing the `-distroless` images](https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/.goreleaser.yml#L11-L12)?_)
- `Dockerfile-alpine` + `Dockerfile-build` are the only relevant images to retain / revise. The [HSM support is minor changes](https://github.com/ory/hydra/pull/3914#issuecomment-2572125066), see TODOs below for future work.
- I've not updated any other `Dockerfile` like the [conformance test](https://github.com/ory/hydra/blob/8e5fae417c19682235cf2f154926a9a3793c1dcd/test/conformance/hydra/Dockerfile), which may warrant adopting the same changes for easier maintenance (_it's effectively `-distroless` variant but with custom cert/CA added, there's better ways to approach that_). Side-note [other `Dockerfile`s for testing](https://github.com/ory/hydra/blob/master/test/e2e/oauth2-client/Dockerfile) are over 5 years old with base images, those should probably get updated.

---

## Future PRs / tasks

Related TODOs from observations while tackling this PR, not to be resolved by this PR:
- [ ] [E2E test in CI is failing](https://github.com/ory/hydra/pull/3914/files#r1900725596) for HSM. This PR doesn't influence that, it's a difference with the CI running Ubuntu while the Docker images use Debian which has the `softhsm` package (_[Issue](https://github.com/ory/hydra/issues/3916)_).
- [ ] DockerHub (_and any other registry that might be published to?_) has misconfigured publishing config (_[Issue](https://github.com/ory/hydra/issues/3917), [Ref](https://github.com/ory/hydra/pull/3914#issuecomment-2571511106)_).
- [ ] Images presently published (v2.2.0) have a dynamically linked hydra binary for Alpine with sqlite support, while the Distroless variant lacks sqlite support (_unless switching to `Dockerfile-build` like `make docker` uses to produce a `-sqlite` image tag_). Both images should be published with the same static binary (_[Ref](https://github.com/ory/hydra/issues/3918), see minor inconsistency section_).
- [ ] DX concerns with docs related to Docker and Development guidance for contributors (_[Issue](https://github.com/ory/hydra/issues/3918), [Issue 2](https://github.com/ory/hydra/issues/3917)_).
- [ ] Follow-up PR should consider dropping non-root, it doesn't seem necessary to maintain, nor properly documented to the image consumer (_[Ref 1](https://github.com/ory/hydra/pull/3914/files#r1903358667), [Ref 2](https://github.com/ory/hydra/pull/3914/files#r1903377234), [Ref 3](https://github.com/ory/hydra/pull/3914/files#r1903378813), [Ref 4](https://github.com/ory/hydra/pull/3914#discussion_r1903355646)_).
- [ ] Follow-up PR could delete a few redundant `Dockerfile` variants (_`Dockerfile-sqlite`, `Dockerfile-scratch`, `Dockerfile-distroless-static`_), which along with defaulting to root as the container user could potentially cover all images in a small single `Dockerfile` (_or two if [keeping HSM separate](https://github.com/ory/hydra/pull/3914#discussion_r1900722821)_) (_[Ref 1](https://github.com/ory/hydra/pull/3914#issuecomment-2572125066). [Ref 2](https://github.com/ory/hydra/pull/3914#discussion_r1903377895), [Ref 3](https://github.com/ory/hydra/pull/3914#issuecomment-2571487713), [Ref 4](https://github.com/ory/hydra/pull/3914#discussion_r1903379223), [Ref 5](https://github.com/ory/hydra/pull/3914#discussion_r1903379260), [Ref 6](https://github.com/ory/hydra/pull/3914#pullrequestreview-2527315326)_).
- [ ] The HSM image final stage is broken. Apparently not used, so could be dropped (_[Ref 1](https://github.com/ory/hydra/pull/3914#issue-2761456550), [Ref 2](https://github.com/ory/hydra/pull/3914#issuecomment-2572125066)_).
- [ ] `GO111MODULE=on` is no longer relevant [since Go 1.16](https://go.dev/blog/go116-module-changes) (Feb 2021), ignored since Go 1.17. You should probably remove [`GO111MODULE` from the various files in this repo](https://github.com/search?q=repo%3Aory%2Fhydra%20GO111MODULE%20&type=code).

File: .docker/Dockerfile-alpine [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
-# By creating the sqlite folder as the ory user, the mounted volume will be owned by ory:ory, which
-# is required for read/write of SQLite.
-RUN mkdir -p /var/lib/sqlite && \
-    chown ory:ory /var/lib/sqlite
+    # Create the sqlite directory with ownership to that user and group:
+    # NOTE: This is required for read/write by SQLite.
+    # - Path may be a default value somewhere, or only explicitly provided via DSN?
+    # - Owner/Group is only relevant to permissions allowing the hydra process to read/write to the location.
+    # - Bind mount volumes will replace the ownership with that of the host directory, requiring correction.
+    install --owner ory --group ory --directory /var/lib/sqlite
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply .docker/Dockerfile-alpine:22]: Nice! The dockerfile_inline would probably need to build the binary inline as well, otherwise the binary type might mismatch (osx versus linux for example)

File: .docker/Dockerfile-alpine [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
+HEREDOC
 
-USER ory
+COPY hydra /usr/bin/hydra
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply .docker/Dockerfile-alpine:25]: Outside of docker compose and `make docker`, the binaries are built by goreleaser and then injected into the containers. So this is correct here.

File: .docker/Dockerfile-alpine [hunk 5]
```diff
@@ -1,22 +1,29 @@
...
 
 ENTRYPOINT ["hydra"]
 CMD ["serve", "all"]
+USER ory
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply .docker/Dockerfile-alpine:29]: I don't agree with this change, the hydra command should run as non-root in my view

File: .docker/Dockerfile-hsm [hunk 6]
```diff
@@ -53,30 +53,36 @@ ENV HSM_LIBRARY=/usr/lib/softhsm/libsofthsm2.so
...
+        --home /home/ory \
+        --ingroup ory \
+        --shell /sbin/nologin \
+        ory
+
+    # Create the sqlite directory with ownership to that user and group:
+    # NOTE: This is required for read/write by SQLite.
+    install --owner ory --group ory --directory /var/lib/sqlite
+
+    # NOTE: Presumably this was already created by the prior RUN directive
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply .docker/Dockerfile-hsm:77]: > Personally unless there's a clear reason for building an image to run with a non-root user, I'd suggest keeping it simple to maintain and just using root which avoids all this.
>
> That leaves the remaining benefit as a "best practice", but AFAIK is mostly moot if the user instead runs the container with Podman/Docker in rootless mode which uses /etc/subuid + /etc/subgid tables on the host to map the containers root UID + GID to the hosts user and any other container UID/GID values to the subuid / subgid range. Alternatively, when run as rootful, the container can be run with all capabilities dropped, which is implicit when running a process as a non-root user in the container.
>
> [...] mostly moot if the user instead  [...]

security is never moot if you assume that everyone is smart or experienced enough to do the right thing. Running as root will fail several CI checks, please revert the user changes made

File: .docker/Dockerfile-scratch [hunk 7]
```diff
@@ -1,28 +1,41 @@
-FROM alpine:3.20
+# TODO: Remove this file in favor of distroless-static variant:
+# https://github.com/ory/hydra/blob/master/.docker/Dockerfile-distroless-static
+# However if published to any registry, continue to publish the variant tag but as an alias to `-distroless` tags:
+# https://github.com/ory/hydra/pull/3914#pullrequestreview-2527315326
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply .docker/Dockerfile-scratch:4]: I checked and we no longer distribute this scratch image but instead the distroless variant. It should be OK to be removed.

**Comment by aeneasr** [comment]
> The HSM image final stage is broken. Apparently not used, so could be dropped (https://github.com/ory/hydra/pull/3914#issue-2761456550, https://github.com/ory/hydra/pull/3914#issuecomment-2572125066).

Unfortunately, that stage is used for CI tests and is causing CI failures

File: .docker/Dockerfile-hsm (hunk 6 above)
[the remark this answers is not in the document]
Comment (polarathene) [reply .docker/Dockerfile-hsm:77]: Uhh, I generally don't see people understanding non-root practice in containers, just pushing for it as a security practice regardless of knowing why.

As I've said, you'll have default capabilities granted to the container root user, but these aren't the same as host root (_full caps_). Switching to a non-root user just drops these additional caps, which if you really need to you can do so yourself as a security practice explicitly (_dropping all caps, or when compatible, switching the user to non-root if you really insist_).

The other reason to use non-root is if container escape is accomplished, then escaping as the root user to be root on the host would allow more damage, but this doesn't make non-root in a container exempt from this, you can still escape as non-root and become root on the host depending on the vulnerability exploited. The bulk of vulnerabilities require granting capabilities or relaxing security intentionally for a container than would otherwise be permitted by default, hence why I'm not really sold on the non-root container user "best practice".

I've seen a variety of containers that adopted non-root as a security practice, but then did workarounds that made security worse. Some granted the non-root user access to the docker socket, others granted capabilities that wouldn't be allowed by default, or mandated capabilities to support non-root but in such a way that prevented someone who is not using a feature that requires that capability and knows how to properly lock down a container to be forced to grant the capability for the binary to run...

To assume that building a container that runs as non-root is always more secure would be a mistake. Inexperienced users that make a container insecure can happen regardless too, trying to address incompetence with a blanket "fix" which often introduces more caveats, especially when the image fails to document this practice is ill-advised IMO. If you're going to defend the practice, have a good reason beyond parroting / hand-waving the justification.

More security conscious (even those inexperienced) should be using rootless Docker/Podman, where it's perfectly ok to have the container user as root, since escaping the container does not result in retaining the same UID/GID of the container user thanks to user namespace mapping.

Root in a container is not equivalent to root on the host. I've explained and demonstrated this so many times in the past to those that don't know better but parrot paranoia insisting that they know what they're talking about.

---

> Running as root will fail several CI checks, please revert the user changes made

I'm not sure what CI checks you were referring to, but your response is to a comment that proposed dropping the non-root approach. The PR respected the non-root decision, all it did here was ensure consistency with the other images for non-root user setup, thus nothing to revert.

File: .docker/Dockerfile-alpine (hunk 5 above)
[the remark this answers is not in the document]
Comment (polarathene) [reply .docker/Dockerfile-alpine:29]: There is no suggestion snippet to remove the line. It was simply shifted to the end of the file. A change to drop it would be:

```suggestion
```

My comment only proposed dropping non-root, and noted that this PR configures the non-root user with a stable UID/GID value, rather than relying on it implicitly (_unstable as any package installed prior may affect user/groups causing the UID/GID for `ory` to change unexpectedly_).

I then noted that the stable UID/GID change itself, would be a breaking change to existing users of the image going forward, similar to if changing from non-root to root.

---

Since there's no real dependency on the non-root user in the image, it's fine and someone can choose to explicitly run as root at runtime (_which should be secure, especially via rootless container_).

File: .docker/Dockerfile-alpine (hunk 3 above)
[the remark this answers is not in the document]
Comment (polarathene) [reply .docker/Dockerfile-alpine:22]: > otherwise the binary type might mismatch (osx versus linux for example)

I don't have macOS environment to verify, but the compose example above is just for a quick example. It copies the binary from the platform image it has, which AFAIK there is no macOS platform for registries/containers, they only publish windows and linux images?

Unless you were referring to architecture like AMD64 vs ARM64? That shouldn't be an issue, the  directive should be using the inferred platform for selecting the image to copy from. Equivalent to `FROM oryd/hydra` or `docker pull oryd/hydra` without `--platform=` specified for either to change the default.

---

The example was focused on showing how much simpler the image is without catering to a non-root user, allowing the deployment to explicitly opt-in to non-root by choosing their own UID/GID, instead of needing to be mindful of what was chosen internally.

That's possible since the image doesn't have any real dependency lock-in on a non-root user being configured. They could just as easily take the current non-root images and run them as root in the same manner.

**Review by polarathene — COMMENTED** [review]
While this PR is now considered redundant as you've raised a replacement PR, I'm wrapping it up with feedback to your last review comments.

No need to respond to the feedback unless you want to engage in that discussion further. I respect your decision to keep non-root if you choose to (_users tend to request it anyway_).

Apologies for the late response, busy week.

**Comment by polarathene** [comment]
> Unfortunately, that stage is used for CI tests and is causing CI failures

If that were the case could you point out where. Pretty sure I identified the CI failure was unrelated to the Docker image or stage.
